app-modules/integration-ai/src/Client/BaseAIChatClient.php

```diff
@@ -0,0 +1,141 @@
+{
+    use InitializesClient;
+
+    protected string $baseEndpoint;
+
+    protected string $apiKey;
+
+    protected string $apiVersion;
+
+    protected string $deployment;
+
+    protected ?string $dynamicContext = null;
+
+    protected ?string $systemContext = null;
+
+    protected Client|ClientFake $client;
+
+    public function __construct(
+    ) {
+        $this->initializeClient();
+    }
+
+    public function ask(Chat $chat, ?Closure $callback): string
+    {
+        if (is_null($this->systemContext)) {
+            $this->setSystemContext();
+        }
+
+        $this->dispatchPromptInitiatedEvent($chat);
+
+        /** @var StreamResponse $stream */
+        $stream = $this->client->chat()->createStreamed([
+            'messages' => $this->formatMessagesFromChat($chat),
+        ]);
+
+        return $this->generateStreamedResponse($stream, $callback);
+    }
+
+    public function provideDynamicContext(DynamicContext $context): self
+    {
+        $this->setDynamicContext($context->context);
+
+        return $this;
+    }
+
+    protected function generateStreamedResponse(StreamResponse $stream, Closure $callback): string
+    {
+        $fullResponse = '';
+
+        foreach ($stream as $response) {
+            $streamedContent = $this->shouldSendResponse($response);
+
+            if (! is_null($streamedContent)) {
+                $callback($streamedContent);
+
+                $fullResponse .= $streamedContent;
+            }
+        }
+
+        return $fullResponse;
+    }
+
+    protected function setSystemContext(): void
+    {
+        $this->systemContext = resolve(AISettings::class)->prompt_system_context;
+    }
+
+    protected function setDynamicContext(string $context): void
+    {
+        $this->dynamicContext = $context;
+    }
+
+    protected function shouldSendResponse(CreateStreamedResponse $response): ?string
+    {
+        if ($response->choices[0]) {
+            $this->examineFinishReason($response);
+        }
+
+        return $response->choices[0]?->delta?->content ?: null;
+    }
+
+    protected function examineFinishReason(CreateStreamedResponse $response): void
+    {
+        match ($response->choices[0]->finishReason) {
+            'length' => throw new TokensExceededException('Your response was not successfully generated due to the max_tokens parameter or token limit being exceeded.'),
+            'content_filter' => throw new ContentFilterException('Your response was not successfully generated due to a flag from our content filters.'),
+            default => null,
+        };
+    }
+
+    protected function formatMessagesFromChat(Chat $chat): array
+    {
+        return [
+            ['role' => 'system', 'content' => $this->addContextToMessages()],
+            ...collect($chat->messages)->map(function (array $message) {
+                return [
+                    'role' => $message['from'],
+                    'content' => $message['message'],
+                ];
```

Comment:
```suggestion
            ...collect($chat->messages)->map(function (ChatMessage $message) {
                return [
                    'role' => $message->from,
                    'content' => $message->message,
                ];
```

At this point, shouldn't the message be the `ChatMessage` DTO? Let's keep the references as such.

Reply:
It should be - but this was resulting in a type error when I originally had that. Will revisit.

Reply:
Should be resolved.